File: tests/python/test_biostatistics_workbench.py
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""Tests for biostatistics-workbench skill scripts.
Covers: glm_pipeline.py
"""
import pytest
import math


def test_response_variable_is_numeric():
    """Response variable must be numeric or integer."""
    response = [12, 5, 0, 8, 3, 15]
    assert all(isinstance(v, (int, float)) for v in response), \
        "Response variable must be numeric"


def test_sample_size_above_minimum():
    """GLM requires at least 20 observations."""
    n_obs = 45
    n_min = 20
    assert n_obs >= n_min, f"Insufficient observations: {n_obs} < {n_min}"


def test_model_family_is_valid():
    """GLM family must be one of the supported families."""
    valid_families = {"gaussian", "poisson", "binomial", "negative_binomial", "gamma"}
    chosen_family = "poisson"
    assert chosen_family in valid_families, f"Unknown GLM family: {chosen_family}"


def test_aic_is_finite():
    """AIC must be a finite number."""
    aic = 234.56
    assert math.isfinite(aic), "AIC must be finite"


def test_effect_sizes_have_ci():
    """Effect size results must include lower and upper 95% CI."""
    result = {
        "estimate": 1.23,
        "ci_lower": 0.85,
        "ci_upper": 1.78
    }
    assert "ci_lower" in result and "ci_upper" in result
    assert result["ci_lower"] < result["estimate"] < result["ci_upper"]


def test_p_value_in_range():
    """p-value must be in [0, 1]."""
    p = 0.032
    assert 0 <= p <= 1, f"p-value out of range: {p}"

File: tests/r/test-biostatistics-workbench.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# test-biostatistics-workbench.R
library(testthat)

DATA_DIR <- if (dir.exists("tests/data")) "tests/data" else file.path("..", "data")

# ── Helper functions matching skill script logic ──────────────────────────────
compute_dispersion_ratio <- function(model) {
  sum(residuals(model, type = "pearson")^2) / df.residual(model)
}

# ── Tests ──────────────────────────────────────────────────────────────────────

test_that("richness data loads with correct structure", {
  df <- read.csv(file.path(DATA_DIR, "richness_data.csv"))
  expect_true("richness"      %in% names(df))
  expect_true("group"         %in% names(df))
  expect_true("elevation"     %in% names(df))
  expect_true("forest_cover"  %in% names(df))
  expect_equal(nrow(df), 90)
})

test_that("three groups are present and balanced in richness data", {
  df <- read.csv(file.path(DATA_DIR, "richness_data.csv"), stringsAsFactors = FALSE)
  counts <- table(df$group)
  expect_equal(as.integer(counts["forest"]),  30)
  expect_equal(as.integer(counts["savanna"]), 30)
  expect_equal(as.integer(counts["pasture"]), 30)
})

test_that("richness values are non-negative integers", {
  df <- read.csv(file.path(DATA_DIR, "richness_data.csv"))
  expect_true(all(df$richness >= 0))
  expect_true(all(df$richness == floor(df$richness)))
})

test_that("Poisson GLM fits without error", {
  df <- read.csv(file.path(DATA_DIR, "richness_data.csv"), stringsAsFactors = FALSE)
  df$group <- factor(df$group)
  m <- glm(richness ~ group + elevation, data = df, family = poisson())
  expect_s3_class(m, "glm")
  expect_false(any(is.na(coef(m))))
})

test_that("forest richness significantly higher than pasture", {
  df <- read.csv(file.path(DATA_DIR, "richness_data.csv"), stringsAsFactors = FALSE)
  forest  <- df$richness[df$group == "forest"]
  pasture <- df$richness[df$group == "pasture"]
  kw <- kruskal.test(list(forest, pasture))
  expect_lt(kw$p.value, 0.05)
  expect_gt(mean(forest), mean(pasture))
})

test_that("dispersion check function works", {
  df <- read.csv(file.path(DATA_DIR, "richness_data.csv"), stringsAsFactors = FALSE)
  df$group <- factor(df$group)
  m <- glm(richness ~ group, data = df, family = poisson())
  disp <- compute_dispersion_ratio(m)
  expect_true(is.numeric(disp))
  expect_gt(disp, 0)
})

test_that("AIC comparison prefers more complex model", {
  df <- read.csv(file.path(DATA_DIR, "richness_data.csv"), stringsAsFactors = FALSE)
  df$group <- factor(df$group)
  m_null <- glm(richness ~ 1,     data = df, family = poisson())
  m_full <- glm(richness ~ group, data = df, family = poisson())
  expect_lt(AIC(m_full), AIC(m_null))
})
